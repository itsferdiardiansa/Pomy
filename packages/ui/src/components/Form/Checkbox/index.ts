export { default as Checkbox } from './Checkbox.vue'
export type { CheckboxTypes } from './props'
