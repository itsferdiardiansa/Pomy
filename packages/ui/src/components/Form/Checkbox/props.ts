export type Item = {
  label: string
}

export type CheckboxTypes = {
  modelValue: string | number | unknown[]
  items: Item[]
  keyname: string
  variant: string
  type: string
}

const checkboxProps = {
  modelValue: '',
  items: () => [],
  keyname: '',
  variant: 'primary',
  type: '',
}

export default checkboxProps
