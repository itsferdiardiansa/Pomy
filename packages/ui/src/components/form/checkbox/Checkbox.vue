<template>
  <div :class="getCls">
    <template v-for="(item, key) in items" :key="key">
      <label :class="`${prefixClass}-ck--item`">
        <input
          type="checkbox"
          v-model="checkboxState"
          :value="getValue(item, key)"
          @change="handleChange"
        />

        <span :class="`${prefixClass}-ck--label`" v-if="item.label">{{
          item.label
        }}</span>
      </label>
    </template>
  </div>
</template>
<script setup lang="ts">
import { computed, unref } from 'vue'
import { prefixClass } from '@/constants'
import defaultProps from './props'
import type { CheckboxTypes } from './props'

const props = withDefaults(defineProps<CheckboxTypes>(), defaultProps)
const emits = defineEmits(['update:modelValue'])

const checkboxState = computed(() => {
  let modelValue = [].concat(props.modelValue)

  return modelValue
})

const getCls = computed(() => {
  const { variant, type } = props
  let customCls = [`${prefixClass}-control--checkbox`]

  if (type === 'button') {
    customCls.push('ck-button')
    customCls.push(`ck-${variant}`)
  }

  return customCls
})

const getValue = (item, key) => {
  const { keyname } = props

  return keyname === undefined ? key : item[keyname]
}

const handleChange = event => {
  const { checked, value } = event.target
  const itemIndex = unref(checkboxState).findIndex(item => item == value)
  let val = [].concat(unref(checkboxState))

  if (checked) val.push(event.target.value)
  else val.splice(itemIndex, 1)

  emits('update:modelValue', val)
}
</script>
<style lang="scss" scoped>
$variants: (
  'primary',
  'danger',
  'warning',
  'dark',
  'success',
  'light',
  'orange'
);

.#{$prefixClass}-ck {
  @each $variant in $variants {
    &-#{$variant} {
      .#{$prefixClass}-ck--item {
        input {
          &:checked + span {
            @apply bg-#{$variant} text-gray-100;
          }
        }
      }

      .#{$prefixClass}-ck--label {
        @apply border border-#{$variant} text-#{$variant};
      }
    }
  }

  &-button {
    .#{$prefixClass}-ck--item {
      input {
        @apply appearance-none;
      }
    }

    .#{$prefixClass}-ck--label {
      @apply border rounded-full py-0.5 px-3 m-0;
    }
  }

  &--item {
    @apply inline-flex items-center mr-2 cursor-pointer my-1;
  }

  &--label {
    @apply ml-2;
  }
}
</style>
