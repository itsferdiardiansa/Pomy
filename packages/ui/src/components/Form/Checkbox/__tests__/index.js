import { mount } from '@vue/test-utils'
import Checkbox from '@common/Form/Checkbox'

describe('common/Form/Checkbox', () => {
  let planets = []
  let Component

  beforeAll(() => {
    const props = {
      items: ['Earth', 'Mars', 'Venus', 'Saturn'],
    }

    Component = mount(<Checkbox v-model={planets} {...props} />)
  })

  it('will render and match snapshot', () => {
    expect(Component.element).toMatchSnapshot()
  })
})
