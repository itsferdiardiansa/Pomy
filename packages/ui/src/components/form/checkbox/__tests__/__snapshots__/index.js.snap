// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Form/Checkbox will render and match snapshot 1`] = `
<div
  class="ss-control--checkbox"
>
  
  <label
    class="ck-item"
  >
    <input
      type="checkbox"
      value="0"
    />
    <!--v-if-->
  </label>
  <label
    class="ck-item"
  >
    <input
      type="checkbox"
      value="1"
    />
    <!--v-if-->
  </label>
  <label
    class="ck-item"
  >
    <input
      type="checkbox"
      value="2"
    />
    <!--v-if-->
  </label>
  <label
    class="ck-item"
  >
    <input
      type="checkbox"
      value="3"
    />
    <!--v-if-->
  </label>
  
</div>
`;
